rfc/0017-error-values/README.md

```diff
@@ -0,0 +1,181 @@
+
+## Explanation
+
+This RFC introduces two error values:
+- `invalid`: Represents an invalid value.
+- `missing`: Represents a missing value.
+
+This distinction is made in order to allow both users and the interpreter more
+fine grained control.
+
+For now, these values are valid for all types (see
+[Possible Future Changes/Enhancements](#possible-future-changesenhancements))
+
+### invalid
+
+This error's primary use case is to represent an erroneous calculation result
+(e.g. Division by zero). It is intended to be used by operator evaluators.
+
+Additionally, parsers can represent a failed value parse using `invalid` (e.g.
+when attempting to parse a number but encountering a letter).
+
+### missing
+
+This error's use case is related to parsing text data and exporting SQL.
+When a parser encounters a missing value (e.g. empty cell in CSV) it can now use
+this error instead of crashing.
+Similarly, SQL exporters can now replace any table cell containing `missing`
+with `NULL`.
+
+### Operator interactions
+
+- Unary operators:
+  - If the parameter is `invalid`, the result is `invalid`
+  - If the parameter is `missing`, the result is `missing`
+  - The following operations evaluate to `invalid`:
+    - Square root of a negative number
+    - Parsing failure for `asDecimal`, `asInteger` or `asBoolean`
+
+- Binary operators:
+  - The following operations always evaluate to `invalid`:
+    - Division by zero
+    - Root of a negative number
+    - 0th root of a number
+    - Number modulo zero
+  - If at least one of the parameters is `invalid`, the result is `invalid`
+  - If at least one of the parameters is `missing` and no parameter is `invalid`, the result is `missing`
+
+- Ternary operators:
+  - If at least one of the parameters is `invalid`, the result is `invalid`
+  - If at least one of the parameters is `missing` and no parameter is `invalid`, the result is `missing`
+
+### Example
+
+data.csv:
+```csv
+column
+3
+```
+
+pipeline.jv:
+```
+pipeline CarsPipeline {
+  Extractor
+    -> ToTextFile
+    -> ToCSV
+    -> ToTable
+    -> ProduceInvalid
+    -> ToSQLite;
+
+  block Extractor oftype LocalFileExtractor {
+    filePath: "data.csv";
+  }
+  block ToTextFile oftype TextFileInterpreter {}
+  block ToCSV oftype CSVInterpreter {}
+  block ToTable oftype TableInterpreter {
+    header: true;
+    columns: [
+      "column" oftype integer,
+    ];
+  }
+  /*
+  | column |
+  |--------|
+  | 3      |
+  */
+  transform divideByZero {
+    from num oftype integer;
+    to out oftype decimal;
+
+    // This calculation results in `invalid`
+    out: num / 0;
+  }
+  block ProduceInvalid oftype TableTransformer {
+    inputColumns: ["column"];
+    outputColumn: "result";
+    uses: divideByZero;
+  }
+  /*
+  | column | result  |
+  |--------|---------|
+  | 3      | invalid |
+  */
+  // `invalid` is replaced with `NULL`
+  block ToSQLite oftype SQLiteLoader {
+    table: "Table";
+    file: "data.sqlite";
+  }
```

Comment:
I am unsure how we should handle the two introduced error values.
Until now, we cannot express something like "number or invalid" in table interpreters.
Should we assume every value can be x, invalid, or missing? Or proceed as we do right now and just remove every row that has invalid or missing values?

@rhazn thoughts?

Reply:
Invalid or missing are correct values for any value type according to this RFC so they will not be removed and instead just be saved as NULL.

Not sure if I have a strong opinion before the RFC is finished, so I'd wait for it to be out of draft before reviewing.